### PR TITLE
Initial oauth implementation

### PR DESCRIPTION
**Pull request by rcypher-databricks**
Oauth implementation.
Added m2m authenticator.
Added basic u2m authenticator.  Implementation is broken out into public functions that clients can use to implement their own authenticator.
Allow specifying auth type and parameters in dsn.

### CODE REVIEW

**Comment by scottlepp** [comment]
Hi @rcypher-databricks this would be a great feature.  I see it's still in draft.  Curious if it's close to merge?

**Comment by rcypher-databricks** [comment]
> Hi @rcypher-databricks this would be a great feature. I see it's still in draft. Curious if it's close to merge?

Hi Scott, the plan is to get back to oauth next week.  So hopefully in the next 2 weeks we'll get something released

File: auth/oauth/oauth.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+	cloudType := InferCloudFromHost(hostName)
+	if cloudType == Azure {
+		userImpersonationScope := fmt.Sprintf("%s/user_impersonation", azureEnterpriseAppId)
+		if !hasScope(scopes, userImpersonationScope) {
+			scopes = append(scopes, userImpersonationScope)
+		}
+	} else {
+		if !hasScope(scopes, "sql") {
+			scopes = append(scopes, "sql")
```
Comment (andrefurlan-db) [inline auth/oauth/oauth.go:60]: should we use sql or all-apis? Gotta check with Bryan Mcquade

File: examples/oauth/main.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	// 	os.Getenv("DATABRICKS_HOST"),
+	// )
+
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+	connector, err := dbsql.NewConnector(
+		dbsql.WithServerHostname(os.Getenv("DATABRICKS_HOST")),
+		dbsql.WithHTTPPath(os.Getenv("DATABRICKS_HTTPPATH")),
+		dbsql.WithDefaultOAUTH(),
```
Comment (andrefurlan-db) [inline examples/oauth/main.go:37]: OAUTH is Oauth

File: internal/config/config.go [hunk 3]
```diff
@@ -14,7 +14,8 @@ import (
 
 	"github.com/databricks/databricks-sql-go/auth"
 	"github.com/databricks/databricks-sql-go/auth/noop"
-	"github.com/databricks/databricks-sql-go/auth/pat"
+	"github.com/databricks/databricks-sql-go/internal/auth/oauth/defauth"
+	"github.com/databricks/databricks-sql-go/internal/auth/pat"
```
Comment (andrefurlan-db) [inline internal/config/config.go:18]: this is a breaking change!

**Review by andrefurlan-db — APPROVED** [review]

**Pull request merged**